### ISSUE TITLE
MLOps-1 — Fine-tune domain-specific LLM on open-source French insurance & regulatory data (MLOps pipeline)

### ISSUE THREAD

**Issue by AndreLiar** [andrelair-platform/platform-backlog#123]
## Context

Generic LLMs (Mistral, LLaMA, Gemma) know insurance concepts in general but do not know:
- French-specific regulatory vocabulary (ACPR, AMF, Code des assurances, DDA/IDD)
- Specific French insurance contract clauses and their legal interpretation
- IBNR reserving methodology in French GAAP context
- Precise answers to "what does Article L112-2 require?" without hallucinating

Fine-tuning a small open-source model on curated French insurance + regulatory open data produces a **domain-specific model** that outperforms GPT-4 on insurance compliance questions while running fully on-premise. This is exactly the kind of specialised AI asset enterprises pay to build — and cannot buy off the shelf.

This issue also adds **MLOps** to the portfolio: dataset curation, training pipeline, evaluation, model registry, versioning, and production serving. That signals ML Engineer / AI Platform Engineer competency directly.

---

## Base model candidates

| Model | Size | Why suitable |
|---|---|---|
| Mistral 7B Instruct v0.3 | 7B | Strong French language, fits on cluster RAM for inference |
| LLaMA 3.1 8B Instruct | 8B | Best open model for instruction following, French support |
| Gemma 2 9B Instruct | 9B | Google quality, efficient inference |
| Mistral Nemo 12B | 12B | Stronger reasoning, multilingual — if swift-mac has enough RAM |

**Recommended**: LLaMA 3.1 8B Instruct — best balance of French language quality, instruction following, and inference cost on the cluster (fits in ~10 GiB RAM via Ollama 4-bit quantisation).

---

## Open-source training data

All data is publicly available, no copyright issues.

### French regulatory corpus

| Source | Content | Volume |
|---|---|---|
| Légifrance (data.gouv.fr) | Code des assurances, Code civil (livres I-IV), Code monétaire et financier | ~50k articles |
| EUR-Lex | Directive Solvabilité II, DORA, DDA/IDD, GDPR — French versions | ~500 documents |
| ACPR publications | Recommandations, Instructions, Décisions publiées (acpr.banque-france.fr) | ~200 documents |
| Journal Officiel | Published insurance-related décrets and arrêtés | Continuous |
| AMF rulebook | Règlement général AMF (French) | ~300 pages |

### Insurance domain knowledge

| Source | Content | Volume |
|---|---|---|
| Wikipedia FR | Assurance, IBNR, bonus-malus, coassurance, réassurance — French articles | ~500 articles |
| OpenAlex (open access) | Academic papers on French insurance, actuarial science, ACPR compliance | ~1000 papers |
| Common Crawl (filtered) | French insurance company FAQs, product pages, glossaries | Filtered subset |
| Légifrance jurisprudence | Court decisions on insurance contract disputes | ~10k decisions |

### Instruction dataset format (SFT — Supervised Fine-Tuning)

Each example is a question/answer pair derived from the regulatory corpus:

```jsonl
{"instruction": "Que prévoit l'article L112-2 du Code des assurances concernant les devis ?",
 "response": "L'article L112-2 impose à l'assureur de remettre à l'assuré un devis avant la conclusion du contrat, précisant les garanties proposées, les exclusions et le montant de la prime. Ce devis doit être conservé et est opposable en cas de litige."}

{"instruction": "Qu'est-ce que la DSN et quelles sont les obligations de l'employeur ?",
 "response": "La Déclaration Sociale Nominative (DSN) est une déclaration mensuelle unique remplaçant la majorité des déclarations sociales. L'employeur doit la transmettre via net-entreprises.fr avant le 5 ou le 15 du mois suivant, selon la taille de l'entreprise, sous peine de pénalités URSSAF."}

{"instruction": "Expliquez le mécanisme du bonus-malus en assurance auto français.",
 "response": "Le coefficient de réduction-majoration (CRM), dit bonus-malus, est régi par l'article A335-9-1 du Code des assurances. Il démarre à 1.00, diminue de 5% par année sans sinistre (plancher 0.50) et augmente de 25% par sinistre responsable (plafond 3.50)..."}
```

Target dataset size: **10,000–50,000 instruction pairs** (sufficient for domain adaptation with QLoRA).

---

### Problem: the cluster has no GPU

The 5 ThinkPad nodes + swift-mac have no dedicated GPU. Fine-tuning even a 7B model on CPU would take weeks.

### Solution: cloud GPU for training, cluster for serving

| Phase | Where | Tool | Cost |
|---|---|---|---|
| Dataset preparation | Mac / cluster (CPU) | Python + HuggingFace datasets | ~0 |
| Fine-tuning (QLoRA) | RunPod / Lambda Labs / Modal | Unsloth + TRL on A100 40GB | ~10–30 € per run |
| Model evaluation | Cluster (CPU inference) | lm-eval-harness | ~0 |
| Production serving | Cluster — Ollama | GGUF 4-bit quantisation | ~0 |

**RunPod A100 40GB**: ~1.5 $/hr. A QLoRA fine-tuning run on 7B model with 10k examples takes ~2–4 hours = ~3–6 $ per training run.

### Training pipeline (Unsloth + TRL)

```python
# train.py — runs on RunPod GPU pod
from unsloth import FastLanguageModel
from trl import SFTTrainer
from datasets import load_dataset

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name="meta-llama/Meta-Llama-3.1-8B-Instruct",
    max_seq_length=2048,
    load_in_4bit=True,
)

model = FastLanguageModel.get_peft_model(
    model,
    r=16,                    # LoRA rank
    target_modules=["q_proj", "v_proj", "k_proj", "o_proj"],
    lora_alpha=16,
    lora_dropout=0,
)

dataset = load_dataset("json", data_files="insurance-fr-sft.jsonl")

trainer = SFTTrainer(
    model=model,
    tokenizer=tokenizer,
    train_dataset=dataset["train"],
    dataset_text_field="text",
    max_seq_length=2048,
    num_train_epochs=3,
    per_device_train_batch_size=2,
    gradient_accumulation_steps=4,
    learning_rate=2e-4,
    output_dir="./insurance-llama-lora",
)
trainer.train()
model.save_pretrained_gguf("insurance-llama-8b", tokenizer, quantization_method="q4_k_m")
```

### Export and serve on cluster

```bash
# Push GGUF to Harbor
crane push insurance-llama-8b.Q4_K_M.gguf harbor.10.0.0.200.nip.io/library/insurance-llm:v1

# Load in Ollama (on cluster via SSH)
ssh controller "kubectl exec -n ai deploy/ollama -- ollama create insurance-fr -f /models/Modelfile"

# Register in LiteLLM gateway
# minicloud-gitops/helm-values/litellm-values.yaml
model_list:
  - model_name: insurance-fr-8b
    litellm_params:
      model: ollama/insurance-fr
      api_base: http://ollama.ai.svc:11434
```

---

### Benchmark: French insurance QA dataset (hand-curated)

100 held-out Q&A pairs across 5 categories:

| Category | Examples |
|---|---|
| Code des assurances | "Quel est le délai de prescription en assurance vie ?" |
| ACPR compliance | "Quelles sont les obligations de reporting ACPR pour un sinistre majeur ?" |
| Claims handling | "Comment calculer une indemnité selon la règle proportionnelle ?" |
| DDA/IDD | "Quelles sont les 4 exigences de la DDA concernant la formation ?" |
| Actuarial | "Définissez le ratio combiné et son interprétation ?" |

### Evaluation metrics

| Metric | Tool | Target |
|---|---|---|
| Exact match on regulatory articles | Custom eval script | > 80% |
| BERTScore FR (semantic similarity) | bert-score | > 0.85 |
| Hallucination rate (article citations) | Spot-check 20 responses | < 5% |
| Latency (time to first token) | Cluster inference test | < 3s on CPU |
| Comparison vs base LLaMA 3.1 8B | lm-eval-harness | +15% accuracy on insurance Q&A |

### Langfuse integration for production monitoring

```python
# Tag all inference calls with model version
langfuse.trace(
    name="insurance-compliance-query",
    metadata={"model": "insurance-fr-8b", "version": "v1.0", "base": "llama3.1-8b"},
    tags=["fine-tuned", "insurance-fr", "compliance"]
)
```

Track fine-tuned model vs base model performance over time in the AI Governance dashboard (#120).

---

## MLOps pipeline (GitOps-managed)

```
Dataset curation (Python scripts)
        ↓
minicloud-gitops/ml-pipeline/
├── datasets/
│   ├── download_legifrance.py
│   ├── download_acpr.py
│   ├── build_sft_dataset.py
│   └── insurance-fr-sft.jsonl   (versioned)
├── training/
│   ├── train.py                  (runs on RunPod)
│   ├── Dockerfile.training
│   └── evaluate.py
└── models/
    ├── Modelfile.insurance-fr    (Ollama model definition)
    └── CHANGELOG.md              (v1.0, v1.1 — what changed)
```

**GitHub Actions workflow:**
```yaml
# .github/workflows/retrain.yml
# Triggered manually or when dataset is updated (PR to datasets/)
on:
  workflow_dispatch:
    inputs:
      runpod_gpu: {type: string, default: "NVIDIA A100 40GB"}
      epochs: {type: number, default: 3}

jobs:
  train:
    # Spin up RunPod pod, run train.py, push GGUF to Harbor, update gitops manifest
```

---

## Portfolio value of this issue

| Skill demonstrated | Who recognises it |
|---|---|
| Dataset curation from open regulatory sources | AI teams at insurers, banks, legal tech |
| QLoRA fine-tuning (Unsloth + TRL) | ML Engineer / AI Engineer hiring managers |
| Cloud GPU orchestration (RunPod) | MLOps / AI Platform roles |
| GGUF quantisation + Ollama serving | On-premise AI deployment specialists |
| Model evaluation and benchmarking | Rigorous ML practice |
| GitOps ML pipeline | MLOps Engineer roles |
| Domain-specific model for insurance compliance | Highly relevant to ACPR-regulated companies |

A fine-tuned insurance model that demonstrably outperforms GPT-4 on French regulatory Q&A is a concrete, testable, differentiating asset. Not many candidates can show this at interview.

---

## Acceptance criteria

- [ ] Dataset pipeline: scripts to download and clean Légifrance + ACPR + EUR-Lex data
- [ ] SFT dataset: 10,000+ instruction pairs in JSONL, versioned in minicloud-gitops
- [ ] Training run completed on RunPod (QLoRA, LLaMA 3.1 8B or Mistral 7B)
- [ ] GGUF model pushed to Harbor (`harbor.10.0.0.200.nip.io/library/insurance-llm:v1`)
- [ ] Model loaded in Ollama on cluster
- [ ] Registered in LiteLLM as `insurance-fr-8b` model route
- [ ] Evaluation: 100-question benchmark, results documented
- [ ] Langfuse traces tag fine-tuned model separately from base models
- [ ] AI Governance dashboard (#120) shows fine-tuned vs base model comparison panel
- [ ] GitHub Actions retrain workflow (manual trigger)
- [ ] Docusaurus page: `docs/ai-platform/fine-tuning-pipeline.md`

## References

- Unsloth: https://github.com/unslothai/unsloth
- TRL (HuggingFace): https://huggingface.co/docs/trl
- Légifrance open data: https://www.data.gouv.fr/fr/datasets/legi-codes-lois-et-reglements-consolides/
- ACPR publications: https://acpr.banque-france.fr/publications
- EUR-Lex French: https://eur-lex.europa.eu/homepage.html?locale=fr
- RunPod: https://www.runpod.io
- Related: #89 (AI Platform EPIC), #120 (AI Governance), #121 (MCP connectors), #122 (document governance)
